src/runtime/Codecs/FunctionCodec.cs

```diff
@@ -0,0 +1,377 @@
+        }
+
+        private static bool IsUnaryAction(Type targetType)
+        {
+            return targetType == typeof(Action);
+        }
+
+        private static bool IsVariadicObjectAction(Type targetType)
+        {
+            return targetType == typeof(Action<object[]>);
+        }
+
+        private static bool IsUnaryFunc(Type targetType)
+        {
+            return targetType == typeof(Func<object>);
+        }
+
+        private static bool IsVariadicObjectFunc(Type targetType)
+        {
+            return targetType == typeof(Func<object[], object>);
+        }
+
+        private static bool IsAction(Type targetType)
+        {
+            return IsUnaryAction(targetType) || IsVariadicObjectAction(targetType);
+        }
+
+        private static bool IsFunc(Type targetType)
+        {
+            return IsUnaryFunc(targetType) || IsVariadicObjectFunc(targetType);
+        }
+
+        private static bool IsCallable(Type targetType)
+        {
+            //TODO - delegate, event, ...
+            return IsAction(targetType) || IsFunc(targetType);
+        }
+
+        public static FunctionCodec Instance { get; } = new FunctionCodec();
+        public bool CanDecode(PyObject objectType, Type targetType)
+        {
+            //python object must be callable
+            if (!objectType.IsCallable()) return false;
+
+            //C# object must be an Action
+            if (!IsCallable(targetType))
+                return false;
+
+            //We don't know if it will work without the instance
+            //The number of arguments of a unary or variadic object callable
+            //is always going to be 0 or 1
+            return true;
+        }
+
+        private static object ConvertUnaryAction(PyObject pyObj)
+        {
+            Func<object> func = (Func<object>)ConvertUnaryFunc(pyObj);
+            Action action = () => { func(); };
+            return (object)action;
+        }
+
+        private static object ConvertVariadicObjectAction(PyObject pyObj, int numArgs)
+        {
+            Func<object[], object> func = (Func<object[], object>)ConvertVariadicObjectFunc(pyObj, numArgs);
+            Action<object[]> action = (object[] args) => { func(args); };
+            return (object)action;
+        }
+
+        //TODO share code between ConvertUnaryFunc and ConvertVariadicObjectFunc
+        private static object ConvertUnaryFunc(PyObject pyObj)
+        {
+            Func<object> func = () =>
+            {
+                Runtime.XIncref(pyObj.Handle);
+                PyObject pyAction = new PyObject(pyObj.Handle);
+                var pyArgs = new PyObject[0];
+                using (Py.GIL())
+                {
+                    var pyResult = pyAction.Invoke(pyArgs);
+                    Runtime.XIncref(pyResult.Handle);
+                    object result;
+                    Converter.ToManaged(pyResult.Handle, typeof(object), out result, true);
```

Comment:
NIT: `out var result`

Reply:
@lostmsu thanks, I didn't know about that